DIPs/DIP1045.md

```diff
@@ -0,0 +1,55 @@
+# Disallow Unsafe Declarations
+
+| Field           | Value                                                           |
+|-----------------|-----------------------------------------------------------------|
+| DIP:            |                                                                 |
+| Review Count:   |                                                                 |
+| Author:         | [Mark](markleaf131313@gmail.com)                                |
+| Implementation: |                                                                 |
+| Status:         | Draft                                                           |
+
+## Abstract
+
+Disallow extern C and C++ function declarations from being marked `@safe` or `@trusted`.
+
+## Contents
+* [Rationale](#rationale)
+* [Prior Work](#prior-work)
+* [Description](#description)
+* [Breaking Changes and Deprecations](#breaking-changes-and-deprecations)
+* [Reference](#reference)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Rationale
+
+Foreign code that is interfaced cannot be guaranteed to be safe. It should always be assumed as unsafe. Especially C declarations as the types used are not mangled into the name of the function. It would be possible to link to a C function with the incorrect parameter types.
```

Comment:
Unless there is a body it is foreign code. If you use C or C++ extern and access the function through a declaration instead of importing it, I don't think that should be a valid use case to consider. You can just import it instead or mark it as a D function and have a different wrapper for C/C++ if you need that.

Reply:
That's not true:

```
// foo.di
extern(C) foo(int i);
```

```
// foo.d
extern(C) foo(int i) {
    // look ma, I have a body here
    return i * 2;
```